### PR TITLE
INDEL realigner cleanup

### LINKED ISSUE #1402: INDEL realigner binary search conditional is flipped
This is a bad one. See https://github.com/bigdatagenomics/adam/blame/master/adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/realignment/RealignIndels.scala#L88-L93. However, what's weird with this is that it should indicate that the realigner is borked, and that isn't consistent with validation data, nor our instrumentation telemetry. I've started to pull in some additional debug infrastructure to capture what's going on at each target in order to suss this out.

### PR DESCRIPTION
**Pull request by fnothaft**
Fixes #1402, and provides many additional improvements to the IR core. Still a WIP, testing on the cluster as we speak.

### CODE REVIEW

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/realignment/RealignIndels.scala [hunk 1]
```diff
@@ -421,20 +435,30 @@ private[read] class RealignIndels(
    * @param read Read to test.
    * @param reference Reference sequence to sweep across.
    * @param qualities Integer sequence of phred scaled base quality scores.
+   * @param originalQuality
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/realignment/RealignIndels.scala]: needs doc comment

File: adam-core/src/main/scala/org/bdgenomics/adam/models/MdTag.scala [hunk 2]
```diff
@@ -208,7 +208,9 @@ object MdTag {
 
           // dirty dancing to recalculate match sets
           for (i <- 0 until cigarElement.getLength) {
-            if (reference(referencePos) == sequence(readPos)) {
+            println(reference + " @ " + referencePos + ", " + sequence + " @ " + readPos)
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/MdTag.scala]: Remove

**Comment by fnothaft** [comment]
Jenkins, retest this please.

**Comment by fnothaft** [comment]
Jenkins, retest this please.

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDD.scala [hunk 3]
```diff
@@ -667,8 +669,18 @@ case class AlignmentRecordRDD(
...
+    optReferenceFile: Option[ReferenceFile] = None,
+    falloff: Int = 0): AlignmentRecordRDD = RealignIndelsInDriver.time {
+    replaceRdd(RealignIndels(rdd,
+      consensusModel,
+      isSorted,
+      maxIndelSize,
+      maxConsensusNumber,
+      lodThreshold,
+      maxTargetSize,
+      maxReadsPerTarget))
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDD.scala]: FIXME

**Comment by fnothaft** [comment]
This is good to go from my end.

**Comment by heuermh** [comment]
Still looks good from my end. Will merge tomorrow unless I hear otherwise. Nice work, @fnothaft!

**Comment by fnothaft** [comment]
Thanks @heuermh! Glad to have things cleaned up.

**Comment by heuermh** [comment]
Don't unclip. Yes, or so I mean no?

**Comment by fnothaft** [comment]
@heuermh think this is good to go now; rerunning a final verification run. Can you give the second commit a looksee? That should be everything new since your last pass.

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/realignment/RealignIndels.scala [hunk 4]
```diff
@@ -225,16 +227,46 @@ private[read] object RealignIndels extends Serializable with Logging {
...
+   */
+  def extractSequenceAndQuality(
+    read: RichAlignmentRecord): (String, Seq[Int]) = {
+
+    // is this read clipped at the start or end? if so, how many bases?
+    val richCigar = new RichCigar(read.samtoolsCigar)
+    val startClipped = richCigar.softClippedBasesAtStart
+    val endClipped = richCigar.softClippedBasesAtEnd
+
+    if (unclipReads ||
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/realignment/RealignIndels.scala]: Getting hard to follow the double negative here.  Maybe I need more sleep.  :)

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/realignment/RealignIndels.scala [hunk 5]
```diff
@@ -348,118 +383,58 @@ private[read] class RealignIndels(
...
+                    // update mdtag and cigar
+                    val newMdTag = MdTag.moveAlignment(r,
+                      newCigar,
+                      reference.drop((newStart - refStart).toInt),
+                      newStart).toString()
+                    builder.setMismatchingPositions(newMdTag)
+                    builder.setOldPosition(r.getStart())
+                    builder.setOldCigar(r.getCigar())
+                    val rec = builder.build()
+                    if (r.getReadName == "H06JUADXX130110:1:1108:12070:36897") {
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/realignment/RealignIndels.scala]: Wot?

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/realignment/RealignIndels.scala [hunk 6]
```diff
@@ -348,118 +383,58 @@ private[read] class RealignIndels(
...
                     }
+                    new RichAlignmentRecord(rec)
                   } else {
                     r
                   }
                 } catch {
                   case t: Throwable => {
-                    log.warn("Realigning read %s failed with %s.".format(r, t))
+                    log.warn("Realigning read %s failed with %s. At:".format(r, t))
+                    val stack = t.getStackTrace()
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/realignment/RealignIndels.scala]: `log.warn(String, Throwable)`?

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/realignment/RealignIndels.scala [hunk 7]
```diff
@@ -500,6 +475,135 @@ private[read] class RealignIndels(
...
+   * @param basesHardClippedAtStart Number of bases aligned as hard clipped
+   *   at the start of the read.
+   * @param basesHardClippedAtEnd Number of bases aligned as hard clipped
+   *   at the end of the read.
+   * @param refStart Reference genome start position for this realignment
+   *   target.
+   * @param remappingIdx The location in the consensus sequence where this
+   *   read has been realigned.
+   * @param consensus The consensus variant to realign against.
+   * @peturn Returns a tuple with the (read start, read end, cigar).
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/realignment/RealignIndels.scala]: `@peturn` &rarr; `@return`

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/realignment/RealignIndels.scala [hunk 8]
```diff
@@ -500,6 +475,135 @@ private[read] class RealignIndels(
...
+        new CigarElement(basesHardClippedAtStart, CigarOperator.H),
+        new CigarElement(basesSoftClippedAtStart, CigarOperator.S),
+        new CigarElement(alignedBases, CigarOperator.M),
+        new CigarElement(basesSoftClippedAtEnd, CigarOperator.S),
+        new CigarElement(basesHardClippedAtEnd, CigarOperator.H))
+        .filter(_.getLength > 0))
+    }
+
+    // if read is fully before or fully after the consensus, then we
+    // have a full match
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/realignment/RealignIndels.scala]: thanks for the comments here

**Comment by fnothaft** [comment]
Squashed down; I say this is good to go from my end.

**Comment by heuermh** [comment]
Thank you, @fnothaft!
